OPENTOK-51006 - 	 Create a Captions API sample app

behei-vonage opened the pull request:
This PR adds a sample app for captions API for the client side.

To test this app:
* Make sure you have a server side running - this can be found at https://github.com/opentok/learning-opentok-node.
* In Basic-Captions/js/config.js, change the URL to `http://localhost:8080`
* In Basic-Captions folder, start a local development server, such as `python3 -m http.server` or whatever you prefer.
* Navigate to `localhost:8000` from two tabs
* Open console logs
* Start the captions by clicking a button and start talking - notice the captions flowing.
* Stop captions from either one of the tabs.
* Notice no errors related to captions (you might see an error but that is related to missing `favicon.ico`) 

Note: the server side sample will be added separately to this repo - https://github.com/opentok/learning-opentok-node.

michaelwhittemore commented on Basic-Captions/README.md:
```diff
@@ -0,0 +1,73 @@
+This sample application shows how to connect to an OpenTok session, publish a stream,
+subscribe to a stream, and publish captions to a session.
+
+> **Note** The demo requires a backend that is running an is able to handle OpenTok captions.
```
typo 'an' to 'and'

michaelwhittemore commented on Basic-Captions/js/app.js:
```diff
@@ -0,0 +1,141 @@
+/* global $ OT SAMPLE_SERVER_BASE_URL */
+
+var apiKey;
```
super nit picky, buy given that this is customer facing shouldn't we use `const` & `let`?

behei-vonage replied:
good point, I just copy-pasted the skeleton from a different sample app. will update 🚀

michaelwhittemore commented:
Has this been tested? It seemed like there was an issue earlier when publishing and subscribing in the same tab?

behei-vonage commented:
> Has this been tested? It seemed like there was an issue earlier when publishing and subscribing in the same tab?

@michaelwhittemore, haven't done testing yet, still need a server component 🙏

michaelwhittemore commented on Basic-Captions/js/app.js:
```diff
@@ -0,0 +1,141 @@
+
+  // Subscribe to a newly created stream
+  session.on('streamCreated', async function streamCreated(event) {
+    var subscriberOptions = {
```
```suggestion
    const subscriberOptions = {
```

michaelwhittemore commented on Basic-Captions/js/app.js:
```diff
@@ -0,0 +1,141 @@
+});
+
+function initializeSession() {
+  var session = OT.initSession(apiKey, sessionId);
```
```suggestion
  const session = OT.initSession(apiKey, sessionId);
```

michaelwhittemore commented on Basic-Captions/js/app.js:
```diff
@@ -0,0 +1,141 @@
+  session.connect(token, function connectCallback(error) {
+    // If the connection is successful, initialize a publisher and publish to the session
+    if (!error) {
+      var publisherOptions = {
```
```suggestion
      const publisherOptions = {
```

michaelwhittemore commented:
> > Has this been tested? It seemed like there was an issue earlier when publishing and subscribing in the same tab?
> 
> @michaelwhittemore, haven't done testing yet, still need a server component 🙏

Makes sense, in that case I'd like to hold off approving until we have a server component to interop with if that's ok. If that's a blocker I can approve now.

behei-vonage commented:
> > > Has this been tested? It seemed like there was an issue earlier when publishing and subscribing in the same tab?
> > 
> > 
> > @michaelwhittemore, haven't done testing yet, still need a server component 🙏
> 
> Makes sense, in that case I'd like to hold off approving until we have a server component to interop with if that's ok. If that's a blocker I can approve now.

that makes sense, let's wait until there is a server component 🙏

devwithzachary commented:
I've also requested @conshus review this sample as this repository is managed by the DevRel team, please let him know when it is ready to rest and review. Thank you!

conshus commented:
Hi all. 
When this is ready, is it cool if I go over it and do some refactoring?
Noticed a couple of things at a first glance:
- In the readme it shows one way of getting the credentials (which is inline with the other web examples), but then app.js does something different.
- This demo relies on the person having a server running instead of just passing in the Credentials
- Could probably remove the jQuery dependency with some small changes
- The index.html links to the Archiving sample

Like I said, if it's cool, once the functionality it complete, I can go in and refactor some things to make it more similar to the other Web Samples.

Thanks.

behei-vonage commented:
> Hi all. When this is ready, is it cool if I go over it and do some refactoring? Noticed a couple of things at a first glance:
> 
> * In the readme it shows one way of getting the credentials (which is inline with the other web examples), but then app.js does something different.
> * This demo relies on the person having a server running instead of just passing in the Credentials
> * Could probably remove the jQuery dependency with some small changes
> * The index.html links to the Archiving sample
> 
> Like I said, if it's cool, once the functionality it complete, I can go in and refactor some things to make it more similar to the other Web Samples.
> 
> Thanks.

Hi @conshus,

Thanks so much for all the suggestions 🙏 I believe that this is ready for review as this sample works with the learning-opentok-node sample app changes. 

If you don't mind helping out with the suggestions you made, I'd greatly appreciate it. 

Thanks!

behei-vonage commented:
@conshus, I believe I addressed most of your feedback on this one. Though, I ran into issues when migrating from ajax to fetch. I was able to get session and token generating to work; however, I keep on getting an empty body when I try to make a request to `/archive/start.` So I kept ajax as is for now, maybe you'd like to give it a try. Thanks!

conshus commented:
@behei-vonage I'll take a look. For the Captions API sample, you are also archiving? I don't see that in the code. Where are you calling `/archive/start`?

behei-vonage commented:
> @behei-vonage I'll take a look. For the Captions API sample, you are also archiving? I don't see that in the code. Where are you calling `/archive/start`?

This sample doesn't call any of the Archiving methods. I updated the index.html to not link to Archiving branch and removed a reference to it in the readme.md file

behei-vonage commented:
We might consider some limitations with this sample app: 

* publisher A joins the session, the captions are started via an Ajax request to the backend, and the client-side captionId is assigned the value that came from the server-side
* publisher B joins the session; it tries to request to start the captions; however, it gets the error 500 with the “captions already enabled” message. The client-side captionId doesn’t get the value from the server
* now, if I try to stop the captions from publisher A, I have no issues. 
* However, if I try to stop them from publisher B, I cannot since the client-side captionsId wasn’t initialized and therefore, the Ajax request from the client to the backend fails.

One way this limitation can be overcome is with the client-side API requesting whether the captions are enabled and grab the captions ID. From discussion on [this thread](https://vonage.slack.com/archives/C0287LFFD0F/p1678143525080249), this might make it in for development in Q2. 

Another way I can think of has a database that stores the captionIds but I feel like it is getting too complex at this point and out of the scope of this Basic-Captions sample app.

cc: @v-kpheng so that you are in the loop and maybe have some suggestions on how to proceed

v-kpheng commented:
> We might need to consider some limitations with this sample app:
> 
> * publisher A joins the session, the captions are started via an Ajax request to the backend, and the client-side captionId is assigned the value that came from the server-side
> * publisher B joins the session; it tries to request to start the captions; however, it gets the error 500 with the “captions already enabled” message. The client-side captionId doesn’t get the value from the server
> * now, if I try to stop the captions from publisher A, I have no issues.
> * However, if I try to stop them from publisher B, I cannot since the client-side captionsId wasn’t initialized and therefore, the Ajax request from the client to the backend fails.
> 
> One way this limitation can be overcome is with the client-side API requesting whether the captions are enabled and grab the captions ID. From discussion on [this thread](https://vonage.slack.com/archives/C0287LFFD0F/p1678143525080249), this might make it in for development in Q2.
> 
> Another way I can think of has a database that stores the captionIds but I feel like it is getting too complex at this point and out of the scope of this Basic-Captions sample app.
> 
> cc: @v-kpheng so that you are in the loop and maybe have some suggestions on how to proceed

Thanks, @behei-vonage.  We can always enhance in Q2, once those APIs are completed.  The sample app you created will help unblock our users.

behei-vonage commented:
awesome @michaelwhittemore helped me out with finding a solution that would work in a 1-to-1 session 💪 

that required some changes to the server side that were done in the following pull request - https://github.com/opentok/learning-opentok-node/pull/40. It'd be great if someone from DevRel could take a look at it. 

@conshus, while it would be great for this sample to look as close to the Archiving samples, there are currently some limitations to what methods we have available. For example, there is a session-level event `archiveStarted`; however, there is not a session-level even `captionsStarted` This method could have been used for retrieving the captionsId before sending a stop request but that is not currently possible; therefore, the stop route on the server side had to be modified to `/captions/stop` instead of /captions/:captionsId/stop`

Please review this sample app when you have a chance and let me know your thoughts. Thanks!

michaelwhittemore commented on Basic-Captions/js/app.js:
```diff
@@ -0,0 +1,133 @@
+      const [subscriberWidget] = subscriberContainer.getElementsByClassName('OT_widget-container');
+    
+      const oldCaptionBox = subscriberWidget.querySelector('.caption-box');
+      if (oldCaptionBox) oldCaptionBox.remove();
```
Super minor nit pick, but we may want to consider breaking this conditional into multiple lines so that it looks cleaner i.e. 
```
if (oldCaptionBox) {
    oldCaptionBox.remove();
}

michaelwhittemore commented on Basic-Captions/README.md:
```diff
@@ -0,0 +1,63 @@
+
+**Token** -- The client also needs a token, which grants them access to the session. Each client is
+issued a unique token when they connect to the session. Since the user publishes an audio-video
+stream to the session, the token generated must include the publish role (the default). For more
```
I think we should mention that starting captions requires a moderator token

michaelwhittemore commented on Basic-Captions/css/app.css:
```diff
@@ -0,0 +1,44 @@
+    left: 20px;
+    z-index: 200;
+}
+.caption-box {
```
I think the caption looked a little small when we were troubleshooting, but I was only seeing it via screenshare so I will leave the judgement call up to you. It's obviously not important or a blocker at all.

behei-vonage replied:
I moved the publisher to the top left instead of the bottom left so that the Publisher object does not block the captions as they come in, but I kept the font size the same; I feel like it's big enough

michaelwhittemore replied:
Awesome, thanks man

michaelwhittemore reviewed (approved):
LGTM, great job

behei-vonage commented:
@conshus could I please have your help in reviewing this one? thanks!

michaelwhittemore commented on Basic-Captions/js/app.js:
```diff
@@ -27,30 +27,26 @@ async function enableTranscription(session) {
+    sessionId,
+    token,
+  }).done(function response() {
+    console.log(captionRes.responseText);
```
Should we include `captions started with ID:`?

behei-vonage commented on Basic-Captions/js/app.js:
```diff
@@ -0,0 +1,132 @@
+  });
+
+  // Connect to the session
+  session.connect(token, async function connectCallback(error) {
```
I don't think it does actually. But this was a copy-pasta from other sample apps so I'll just keep it as-is

behei-vonage commented:
Note: there were a lot of awesome updates by @conshus that were carried over to this PR from his branch - https://github.com/opentok/opentok-web-samples/tree/opentok-51006-refactor/Basic-Captions. I couldn't think of a faster way to bring those changes over here except for copying it over, but the credit for that goes over to Dwane. Thanks for helping out with this 💪

conshus commented:
No problem @behei-vonage . Glad I could help.  Great job! 👍🏾

dwivedisachin commented:
Tested LGTM!! 🚀🚀

dwivedisachin commented:
Tested LGTM!! 🚀🚀
